### PR TITLE
[fnf#13] Classifications - prep 1

### PR DESCRIPTION
**Pull request by gbp**
## Relevant issue(s)

Connect to https://github.com/mysociety/transparency-fnf-whatdotheyknow-projects/issues/13

## What does this do?

Extract existing classification actions into new controllers and start to unify common parts.

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -141,11 +135,18 @@
...
 
-  #### Citations controller
   scope path: 'request/:url_title' do
+    #### Citations controller
     resources :citations, only: [:new, :create]
+    ####
+
+    #### Classifications controller
+    resources :classifications, only: [:create], param: :described_state do
+      get :message, on: :member
```
Comment (gbp) [inline config/routes.rb:145]: @garethrees this generates the route as `/alaveteli_pro/request/:url_title/classifications/:described_state/message` I'm tempted to switch to `/alaveteli_pro/request/:url_title/classifications/:described_state` using the `new` action - thoughts?

File: config/routes.rb [hunk 2]
```diff
@@ -699,9 +700,12 @@
     namespace :alaveteli_pro do
       root to: 'dashboard#index', :as => :dashboard, :via => :get
       resources :draft_info_requests, :only => [:create, :update]
-      resources :info_requests, :only => [:new, :create, :update, :index] do
+      resources :info_requests, only: [:new, :create, :index] do
         get :preview, on: :new # /info_request/new/preview
       end
+      scope path: 'request/:url_title' do
```
Comment (gbp) [inline config/routes.rb:706]:
```suggestion
      scope path: 'info_requests/:url_title' do
```

Perhaps move this block down the file to the other Pro route scope.

File: app/controllers/request_controller.rb [hunk 3]
```diff
@@ -22,13 +22,15 @@ class RequestController < ApplicationController
...
 
+  # rubocop:disable Style/ClassVars, Lint/HandleExceptions
   @@custom_states_loaded = false
   begin
     require 'customstates'
     include RequestControllerCustomStates
     @@custom_states_loaded = true
   rescue LoadError, NameError
   end
+  # rubocop:enable Style/ClassVars, Lint/HandleExceptions
```
Comment (gbp) [inline app/controllers/request_controller.rb:34]: This isn't needed here anymore.

```suggestion
```

File: spec/controllers/alaveteli_pro/classifications_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    let(:other_pro_user) { FactoryBot.create(:pro_user) }
+    let(:info_request) { FactoryBot.create(:info_request, user: pro_user) }
+
+    context 'when the user is not allowed to update the request' do
+      it 'raises a CanCan::AccessDenied error' do
+        session[:user_id] = other_pro_user.id
+        expect do
+          post :create, params: {
+            url_title: info_request.url_title,
+            info_request: { described_state: 'successful' }
```
Comment (gbp) [inline spec/controllers/alaveteli_pro/classifications_controller_spec.rb:15]:
```suggestion
            classification: { described_state: 'successful' }
```

File: config/routes.rb (hunk 1 above)
Comment (garethrees) [reply config/routes.rb:145]: Lets keep it as it is for now – I'm not necessarily sure it should be part of this controller in the first place, but IDK. Will look at that with fresh eyes once we've made the more general restructure.

File: app/controllers/classifications_controller.rb [hunk 5]
```diff
@@ -150,4 +133,22 @@ def message
...
+    @info_request = InfoRequest.not_embargoed.find_by!(
+      url_title: params[:url_title]
+    )
+  end
+
+  def authorise_info_request
+    # If this is an external request, go to the request page - we don't allow
+    # state change from the front end interface.
+    if @info_request.is_external?
+      redirect_to_info_request
```
Comment (garethrees) [inline app/controllers/classifications_controller.rb:148]: Looks like this introduces a spec failure:

```
Failures:

  1) ClassificationsController POST create if the request is external should redirect to the request page
     Failure/Error: redirect_to_info_request

     NameError:
       undefined local variable or method `redirect_to_info_request' for #<ClassificationsController:0x00005623971a5548>
     # ./app/controllers/classifications_controller.rb:149:in `authorise_info_request'
     # /home/vagrant/bundle/ruby/2.5.0/gems/activesupport-5.1.7/lib/active_support/callbacks.rb:413:in `block in make_lambda'
     # /home/vagrant/bundle/ruby/2.5.0/gems/activesupport-5.1.7/lib/active_support/callbacks.rb:197:in `block (2 levels) in halting'
     # /home/vagrant/bundle/ruby/2.5.0/gems/actionpack-5.1.7/lib/abstract_controller/callbacks.rb:12:in `block (2 levels) in <module:Callbacks>'
     # /home/vagrant/bundle/ruby/2.5.0/gems/activesupport-5.1.7/lib/active_support/callbacks.rb:198:in `block in halting'
     # /home/vagrant/bundle/ruby/2.5.0/gems/activesupport-5.1.7/lib/active_support/callbacks.rb:507:in `block in invoke_before'
     # /home/vagrant/bundle/ruby/2.5.0/gems/activesupport-5.1.7/lib/active_support/callbacks.rb:507:in `each'
     # /home/vagrant/bundle/ruby/2.5.0/gems/activesupport-5.1.7/lib/active_support/callbacks.rb:507:in `invoke_before'
     # /home/vagrant/bundle/ruby/2.5.0/gems/activesupport-5.1.7/lib/active_support/callbacks.rb:106:in `block in run_callbacks'
     # ./app/controllers/application_controller.rb:116:in `record_memory'
     # /home/vagrant/bundle/ruby/2.5.0/gems/activesupport-5.1.7/lib/active_support/callbacks.rb:117:in `block in run_callbacks'
     # /home/vagrant/bundle/ruby/2.5.0/gems/activesupport-5.1.7/lib/active_support/callbacks.rb:135:in `run_callbacks'
     # /home/vagrant/bundle/ruby/2.5.0/gems/actionpack-5.1.7/lib/abstract_controller/callbacks.rb:19:in `process_action'
     # /home/vagrant/bundle/ruby/2.5.0/gems/actionpack-5.1.7/lib/action_controller/metal/rescue.rb:20:in `process_action'
     # /home/vagrant/bundle/ruby/2.5.0/gems/actionpack-5.1.7/lib/action_controller/metal/instrumentation.rb:32:in `block in process_action'
     # /home/vagrant/bundle/ruby/2.5.0/gems/activesupport-5.1.7/lib/active_support/notifications.rb:166:in `block in instrument'
     # /home/vagrant/bundle/ruby/2.5.0/gems/activesupport-5.1.7/lib/active_support/notifications/instrumenter.rb:21:in `instrument'
     # /home/vagrant/bundle/ruby/2.5.0/gems/activesupport-5.1.7/lib/active_support/notifications.rb:166:in `instrument'
     # /home/vagrant/bundle/ruby/2.5.0/gems/actionpack-5.1.7/lib/action_controller/metal/instrumentation.rb:30:in `process_action'
     # /home/vagrant/bundle/ruby/2.5.0/gems/actionpack-5.1.7/lib/action_controller/metal/params_wrapper.rb:252:in `process_action'
     # /home/vagrant/bundle/ruby/2.5.0/gems/activerecord-5.1.7/lib/active_record/railties/controller_runtime.rb:22:in `process_action'
     # /home/vagrant/bundle/ruby/2.5.0/gems/actionpack-5.1.7/lib/abstract_controller/base.rb:124:in `process'
     # /home/vagrant/bundle/ruby/2.5.0/gems/actionview-5.1.7/lib/action_view/rendering.rb:30:in `process'
     # /home/vagrant/bundle/ruby/2.5.0/gems/actionpack-5.1.7/lib/action_controller/metal.rb:189:in `dispatch'
     # /home/vagrant/bundle/ruby/2.5.0/gems/actionpack-5.1.7/lib/action_controller/test_case.rb:513:in `process'
     # /home/vagrant/bundle/ruby/2.5.0/gems/rails-controller-testing-1.0.2/lib/rails/controller/testing/template_assertions.rb:61:in `process'
     # /home/vagrant/bundle/ruby/2.5.0/gems/actionpack-5.1.7/lib/action_controller/test_case.rb:400:in `post'
     # /home/vagrant/bundle/ruby/2.5.0/gems/rails-controller-testing-1.0.2/lib/rails/controller/testing/integration.rb:12:in `block (2 levels) in <module:Integration>'
     # ./spec/controllers/classifications_controller_spec.rb:9:in `block (4 levels) in <top (required)>'

Finished in 2 minutes 4.3 seconds (files took 28.47 seconds to load)
79 examples, 1 failure

Failed examples:

rspec ./spec/controllers/classifications_controller_spec.rb:8 # ClassificationsController POST create if the request is external should redirect to the request page
```

File: spec/controllers/alaveteli_pro/classifications_controller_spec.rb (hunk 4 above)
Comment (gbp) [reply spec/controllers/alaveteli_pro/classifications_controller_spec.rb:15]: This should be fixed up as part of bf6702a00d27b0a6a21eb6f531fb9b9732e1302e

File: app/controllers/classifications_controller.rb (hunk 5 above)
Comment (gbp) [reply app/controllers/classifications_controller.rb:148]: Ah yes, that was added in the wrong commit - I've pushed two fixups to remove here and add to the right commit.

File: app/controllers/request_controller.rb (hunk 3 above)
Comment (gbp) [reply app/controllers/request_controller.rb]: Push fixup to remove

File: spec/controllers/alaveteli_pro/classifications_controller_spec.rb (hunk 4 above)
Comment (gbp) [reply spec/controllers/alaveteli_pro/classifications_controller_spec.rb]: Fixup pushed

**Pull request merged**